### ISSUE TITLE
Allow restRequest to be cleanly canceled

### ISSUE THREAD

**Issue by brianhelba** [girder/girder#2445]
While there is a `cancelRestRequests` method in Girder, it should typically not be used, as it cancels all rest requests globally, instead of only those from a specific call.

Instead, a mechanism should be exposed to cancel specific obsolete rest requests when necessary. While the `jqXHR` object returned by `restRequest` does expose an `.abort` method, this is not accessible through any subsequent promise chains. A good comprehensive solution may involve a switch to Bluebird promises, which have [a `.cancel` method](http://bluebirdjs.com/docs/api/cancellation.html), which can be hooked up to the `.abort ` method inside `restRequest`.

In particular, this functionality would be useful to make the [slice slider in `dicom_preview`](https://github.com/girder/girder/blob/a7c24a254467305acbc3ceb5f3527cdf820fe9af/plugins/dicom_viewer/web_client/views/DicomView.js#L334) cancel any still-pending slice fetches when it is changed.

Once this functionality is in-place, `cancelRestRequests` should be deprecated and removed.

**Comment by jbeezley** [comment]
I agree `cancelRestRequests` should be deprecated, but I'm hesitant to introduce any sort of cancelable promises even if it's through bluebird.  There have been a number of attempts for several years to introduce cancelable promises into the ES standard (https://github.com/tc39/proposal-cancelable-promises).  None have made it into the standard for a variety of reasons, but one of the primary reasons AFAIK is that it adds all sorts of difficult edge cases and foot guns.

The utility of "canceling" rest requests is limited anyway because you can't actually cancel the request once the ajax call has been made, you can only ignore the response.  I think it is most sensible to handle on a case-by-case basis to avoid the complexity of the generalization.

**Comment by brianhelba** [comment]
@jbeezley +1 to all.

However, there will still be cases (e.g. the `dicom_viewer`) where a slow connection could cause dozens of unsent requests to pile up. In those cases, it may be possible to totally cancel a request if the actual TCP/HTTP request has not yet been sent (though I'm not sure, and I need to do more research into how browsers throttle multiple Ajax requests), or it may still be useful to ignore responses to prevent strange user-facing side effects from the resolution of "old" requests.

Ultimately, I see this as a being a niche feature which [should only be invoked when it's shown to be necessary](http://wiki.c2.com/?PrematureOptimization). In those cases, it'd be helpful to have some reasonable technical mechanism.

**Comment by jbeezley** [comment]
We went through this exploration in geojs.  @manthey can probably fill in more details, but my recollection is that once the ajax js call is made, there is no way to cancel it even if the browser has not actually sent the request yet.  There is a browser-specific limit to the number of simultaneous requests to a single domain, so additional requests just set in a queue.  The only way we found to handle this is to create our own request queue.  The code for this is in [fetchQueue.js](https://github.com/OpenGeoscience/geojs/blob/master/src/fetchQueue.js) if you are interested.
